libp2p/network/url_conversion.go

```diff
@@ -0,0 +1,133 @@
+package network
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+
+	"github.com/multiformats/go-multiaddr"
+)
+
+const (
+	transportTypeTCP = "tcp"
+	ipVersion4       = "ip4"
+	ipVersion6       = "ip6"
+
+	// anyScheme is a string which fills the "scheme" degree of freedom for
+	// a URL. Used to pad URLs without schemes such that they can be parsed
+	// via stdlib URL parser.
+	anyScheme           = "scheme://"
+	errResolvePeerIPMsg = "resolving peer IP for hostname: %s: %w"
+)
+
+// Libp2pMultiaddrFromServiceUrl transforms a URL into its libp2p multiaddr equivalent.
+// (see: https://github.com/libp2p/specs/blob/master/addressing/README.md#multiaddr-basics)
+func Libp2pMultiaddrFromServiceUrl(serviceUrl string) (multiaddr.Multiaddr, error) {
+	var (
+		// TECHDEBT: assuming TCP; remote peer's transport type must be knowable!
+		// (ubiquitously switching to multiaddr, instead of a URL, would resolve this)
+		peerTransportStr = transportTypeTCP
+		// Default to IPv4; updated via subsequent checks.
+		peerIPVersionStr = ipVersion4
+	)
+
+	peerUrl, err := url.Parse(anyScheme + serviceUrl)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"parsing peer service URL: %s: %w",
+			serviceUrl,
+			err,
+		)
+	}
+
+	peerIP, err := getPeerIP(peerUrl.Hostname())
+	if err != nil {
+		return nil, err
+	}
+
+	// Check IP version.
+	if peerIP.To4() == nil {
+		peerIPVersionStr = "ip6"
+	}
+
+	peerMultiAddrStr := fmt.Sprintf(
+		"/%s/%s/%s/%s",
+		peerIPVersionStr,
+		peerIP,
+		peerTransportStr,
+		peerUrl.Port(),
+	)
+	return multiaddr.NewMultiaddr(peerMultiAddrStr)
+}
+
+// ServiceUrlFromLibp2pMultiaddr converts a multiaddr into a URL string.
+func ServiceUrlFromLibp2pMultiaddr(addr multiaddr.Multiaddr) (string, error) {
+	protocols := addr.Protocols()
+	if len(protocols) < 2 {
+		return "", fmt.Errorf(
+			"unsupported multiaddr: %s; expected at least 2 protocols, got: %d",
+			addr, len(protocols),
+		)
+	}
+
+	networkProtocol := protocols[0]
+	// e.g. IP address: "/ip4/10.0.0.1" --> "10.0.0.1"
+	networkValue, err := addr.ValueForProtocol(networkProtocol.Code)
+	if err != nil {
+		return "", err
+	}
+
+	transportProtocol := protocols[1]
+	// e.g. Port: "/tcp/42069" --> "42069"
+	transportValue, err := addr.ValueForProtocol(transportProtocol.Code)
+	if err != nil {
+		return "", err
+	}
+
+	// Top level protocol must be a network protocol (e.g. ip4, ip6).
+	switch networkProtocol.Code {
+	case multiaddr.P_IP4, multiaddr.P_IP6:
+		return fmt.Sprintf("%s:%s", networkValue, transportValue), nil
+	}
+
+	return "", fmt.Errorf(
+		"unsupported network protocol, %s",
+		networkProtocol.Name,
+	)
+}
+
+func newResolvePeerIPErr(hostname string, err error) error {
+	return fmt.Errorf(errResolvePeerIPMsg, hostname, err)
+}
+
+func getPeerIP(hostname string) (net.IP, error) {
```

Comment:
@okdas, @Olshansk thanks for taking a look, for adding more context, and the advice! :pray: :raised_hands: 

> Would it be a good idea to detect ipv6/ipv6 addresses and skip DNS resolve for them? And only perform DNS request on hostnames.

Yes, absolutely! I've factored out the IP resolution portion to this helper, `getPeerIP` which does this. Additionally, it's now using [`net.LookupHost`](https://pkg.go.dev/net@go1.19.5#LookupHost) (instead of [`net.ResolveIPAddr`](https://pkg.go.dev/net@go1.19.5#ResolveIPAddr)) for DNS resolution which puts us the driver seat with respect to what we want to do with a DNS response containing multiple addresses. I added a few `CONSIDER`s about this (lines 121-123).

Reply:
> which puts us the driver seat with respect to DNS responses

Wdym by this?

Reply:
I meant that we get a list of all addresses in the response and get to decide what to do with it, as opposed to being at the whim of [`net.ResolveIPAddr`](https://pkg.go.dev/net#ResolveIPAddr).

> The address parameter can use a host name, but this is not recommended, because it will return at most one of the host name's IP addresses.

Reply:
EDIT: if fixed my typo, "... puts us **in** the driver seat ...". Sorry for the confusion.

Reply:
Makes more sense. Ty! 

Any chance you can add a small comment on why we're using `net.ResolveIPAddr` instead of `netlLookupHost` directly in the code?

Reply:
Though having multiple A DNS records for the same hostname should not happen often in our case, I can see a scenario when nodes are provisioned behind a multi-zone LB, which is often a default for Kubernetes set up in clouds (I suppose they make more 🤑 that way). 

Do you think there might be issues when the same node has a hostname that resolves to multiple IP addresses?

I assume PeerID has no correlation with Multiaddr. I suppose it should be fine if one PeerID has different addresses. Imagine a situation with multiple network interfaces, e.g. internet and local networks on one machine, both interfaces have different addresses in that case. But the same node can be exposed on both interfaces. In that case, (I assume) the PeerID stays the same, but other peers should be able to connect to it just fine even if IP addresses are different. Not sure if that makes sense, I'm just kind of fascinated by this topic and my guess is random address should work as long as packets flow correctly. :)

Reply:
> puts us in the driver seat

Do you mean something like we could sort the IP addresses and always use the first one so all nodes in the network predictably connect to the same IP even if there are multiple A DNS records?

Reply:
@okdas 

> Do you think there might be issues when the same node has a hostname that resolves to multiple IP addresses?

I think it depends on some things, I think [the canvas](https://github.com/pokt-network/pocket/pull/534#discussion_r1119808016) I put together tries to get at this.

> I assume PeerID has no correlation with Multiaddr. 

Correct, mapping from libp2p to pocket nomenclature: `PeerID` is analagous to `Address`, and `Multiaddr` to `ServiceURL`.

> I suppose it should be fine if one PeerID has different addresses.

Yes, I believe this is the case. Libp2p's [`peerstore.Peerstore`](https://pkg.go.dev/github.com/libp2p/go-libp2p@v0.26.1/core/peerstore#Peerstore) interface embeds [`peerstore.AddrBook`](https://pkg.go.dev/github.com/libp2p/go-libp2p@v0.26.1/core/peerstore#AddrBook) which makes it clear via its API that multiple multiaddrs can be associated with a given `PeerID`.

> Not sure if that makes sense, I'm just kind of fascinated by this topic and my guess is random address should work as long as packets flow correctly. :)

Indeed, that does make sense, and I agree! You've also actually called out something that I didn't completely think through, which is randomizing the selected DNS record. I was just selecting the first valid record which I'm realizing would not serve the high availability goal. :facepalm: :sweat_smile: 53067bd should address this.

> Do you mean something like we could sort the IP addresses and always use the first one so all nodes in the network predictably connect to the same IP even if there are multiple A DNS records?

Given the above, I would question the value in each node having the same topological view of the overlay network.

I just meant that we have control as opposed to being limited by [`net.ResolveIPAddr`](https://pkg.go.dev/net#ResolveIPAddr), which was the case before. It would only return a single record given a hostname:

> ... The address parameter can use a host name, but this is not recommended, because it will return at most one of the host name's IP addresses.